xmlSignature + business logic queries implementation

zanni opened the pull request:
Bryan,

Here are the changes we discussed before. This pull request contains:
- xmlSignature verify
- compatibility with ECDSA P-384 keys
- CreateReport / CancelReport queries
- PartyReRegistrer / CancelPartyRegistration query

I implemented the  verify function the same way I implemented the sign function. Therefore, I had to "promisify" the message reading logic (prepareResMsg, prepareReqMsg) a bit.

I also implemented functions to manipulate signatureWrapper you can use (hasWrapper, getWrappedPayloadName, wrap, unwrap). Have a look at prepareResMsg, prepareReqMsg in order to see how to use them.

I modified a bit your demo flow in order to implements oadr business logic on previously mentioned queries. You can found this modified flow in example-flows/demoflow_pullrequest.json file.

Best regards
Bertrand Zanni

bnystrom commented:
I've pulled your code into branch "pull/No4" and done some changes:


   - Re-format and make cosmetic changes to satisfy the linter (please lint)
   - Messages w/o XML signature default to adding the
   oadrPayload/oadrSignedObject env wrappers
   - Added an option in the config dialog to NOT do the above (in case
   that's the way a particular VTN likes it)
   - Always remove the "TYPE_NAME" property that is only needed by jsonix
   transforms when outputing to Node-Red flows.
   - Removed an extraneous extra call to the model_builder which existed in
   the CreatePartyRegistraion function (I left it commented out in case you
   have a valid reason for keeping it that I'm not aware of)

Go ahead verify with that branch if I've broken anything on your end. I'll
do a bit more testing and if all is well, I'll release soon.


Best Regards
Bryan Nystrom


On Wed, Mar 13, 2019 at 11:14 AM zanni bertrand <notifications@github.com>
wrote:

> Bryan,
>
> Here are the changes we discussed before. This pull request contains:
>
>    - xmlSignature verify
>    - compatibility with ECDSA P-384 keys
>    - CreateReport / CancelReport queries
>    - PartyReRegistrer / CancelPartyRegistration query
>
> I implemented the verify function the same way I implemented the sign
> function. Therefore, I had to "promisify" the message reading logic
> (prepareResMsg, prepareReqMsg) a bit.
>
> I also implemented functions to manipulate signatureWrapper you can use
> (hasWrapper, getWrappedPayloadName, wrap, unwrap). Have a look at
> prepareResMsg, prepareReqMsg in order to see how to use them.
>
> I modified a bit your demo flow in order to implements oadr business logic
> on previously mentioned queries. You can found this modified flow in
> example-flows/demoflow_pullrequest.json file.
>
> Best regards
> Bertrand Zanni
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>
> https://github.com/Argonne-National-Laboratory/node-red-contrib-oadr-ven/pull/4
> Commit Summary
>
>    - oadr20b model generation using jsonix
>    - cleanup
>    - xmldsig
>    - xmldsig
>    - xmldsig
>    - cleanup + demo flow
>    - profil flow variable
>    - createOpt fix
>    - xmldsig
>    - canceledPartyRegistration response
>    - handle createdReport canceledReport
>    - xmldsig verify + xmldsig ECDSA P-384 + CreateReport + CancelReport +
>    ReRegistration
>
> File Changes
>
>    - *M* .gitignore
>    <https://github.com/Argonne-National-Laboratory/node-red-contrib-oadr-ven/pull/4/files#diff-0>
>    (2)
>    - *A* example-flows/demoflow_pullrequest.json
>    <https://github.com/Argonne-National-Laboratory/node-red-contrib-oadr-ven/pull/4/files#diff-1>
>    (1)
>    - *M* oadr2/oadr2-ven.js
>    <https://github.com/Argonne-National-Laboratory/node-red-contrib-oadr-ven/pull/4/files#diff-2>
>    (977)
>    - *A* oadr2/oadr2b-payload.js
>    <https://github.com/Argonne-National-Laboratory/node-red-contrib-oadr-ven/pull/4/files#diff-3>
>    (358)
>    - *A* oadr2/oadr2b_model/oadr2b_model.js
>    <https://github.com/Argonne-National-Laboratory/node-red-contrib-oadr-ven/pull/4/files#diff-4>
>    (6927)
>    - *A* oadr2/oadr2b_model/oadr2b_model.jsonschema
>    <https://github.com/Argonne-National-Laboratory/node-red-contrib-oadr-ven/pull/4/files#diff-5>
>    (22637)
>    - *D* package-lock.json
>    <https://github.com/Argonne-National-Laboratory/node-red-contrib-oadr-ven/pull/4/files#diff-6>
>    (1696)
>    - *M* package.json
>    <https://github.com/Argonne-National-Laboratory/node-red-contrib-oadr-ven/pull/4/files#diff-7>
>    (7)
>
> Patch Links:
>
>    -
>    https://github.com/Argonne-National-Laboratory/node-red-contrib-oadr-ven/pull/4.patch
>    -
>    https://github.com/Argonne-National-Laboratory/node-red-contrib-oadr-ven/pull/4.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Argonne-National-Laboratory/node-red-contrib-oadr-ven/pull/4>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABsML2VqFXFvTysi9AreKoxjlPHg7bKvks5vWSPPgaJpZM4btoqb>
> .
>

zanni commented:
I merged your code into my fork and changed few things:

- I had to uncomment the call to model_builder in
oadrCreatePartyRegistration because of xmlSignature config. Indeed the only
way to configure xmlSignature profile is by adding "xmlSignature:true" to
createPartyRegistration payload. the line:
xmlSignature = config.xmlSignature || false;
at node-red module initialization is useless because there is no
xmlSignature configuration at node level (node-red UI).

- I added a proper oadr response when a oadrDistributeEvent payload is
pushed from VTN to VEN

- I changed the createdEvent function to support pull/push profile

- I changed "responseCode" default value in different payload because this
variable is supposed to be a string

- I removed an unused dependency from package.json (node-webcrypto-p11)

- I executed eslint over the code using this command: eslint ./oadr2/ --fix
which seems to have done what I expected (code indent etc...). However I'm
not used to eslint so don't hesitate to specify the workflow you prefer
regarding that point. Maybe this should be scripted as "npm run" cmd in
package.json file.

Do you want me to make a pull request of those changes into the develop
branch of your repo ?

Best regards,

Bertrand

Le lun. 18 mars 2019 à 22:50, Bryan Nystrom <notifications@github.com> a
écrit :

> I've pulled your code into branch "pull/No4" and done some changes:
>
>
> - Re-format and make cosmetic changes to satisfy the linter (please lint)
> - Messages w/o XML signature default to adding the
> oadrPayload/oadrSignedObject env wrappers
> - Added an option in the config dialog to NOT do the above (in case
> that's the way a particular VTN likes it)
> - Always remove the "TYPE_NAME" property that is only needed by jsonix
> transforms when outputing to Node-Red flows.
> - Removed an extraneous extra call to the model_builder which existed in
> the CreatePartyRegistraion function (I left it commented out in case you
> have a valid reason for keeping it that I'm not aware of)
>
> Go ahead verify with that branch if I've broken anything on your end. I'll
> do a bit more testing and if all is well, I'll release soon.
>
>
> Best Regards
> Bryan Nystrom
>
>
> On Wed, Mar 13, 2019 at 11:14 AM zanni bertrand <notifications@github.com>
> wrote:
>
> > Bryan,
> >
> > Here are the changes we discussed before. This pull request contains:
> >
> > - xmlSignature verify
> > - compatibility with ECDSA P-384 keys
> > - CreateReport / CancelReport queries
> > - PartyReRegistrer / CancelPartyRegistration query
> >
> > I implemented the verify function the same way I implemented the sign
> > function. Therefore, I had to "promisify" the message reading logic
> > (prepareResMsg, prepareReqMsg) a bit.
> >
> > I also implemented functions to manipulate signatureWrapper you can use
> > (hasWrapper, getWrappedPayloadName, wrap, unwrap). Have a look at
> > prepareResMsg, prepareReqMsg in order to see how to use them.
> >
> > I modified a bit your demo flow in order to implements oadr business
> logic
> > on previously mentioned queries. You can found this modified flow in
> > example-flows/demoflow_pullrequest.json file.
> >
> > Best regards
> > Bertrand Zanni
> > ------------------------------
> > You can view, comment on, or merge this pull request online at:
> >
> >
> >
> https://github.com/Argonne-National-Laboratory/node-red-contrib-oadr-ven/pull/4
> > Commit Summary
> >
> > - oadr20b model generation using jsonix
> > - cleanup
> > - xmldsig
> > - xmldsig
> > - xmldsig
> > - cleanup + demo flow
> > - profil flow variable
> > - createOpt fix
> > - xmldsig
> > - canceledPartyRegistration response
> > - handle createdReport canceledReport
> > - xmldsig verify + xmldsig ECDSA P-384 + CreateReport + CancelReport +
> > ReRegistration
> >
> > File Changes
> >
> > - *M* .gitignore
> > <
> https://github.com/Argonne-National-Laboratory/node-red-contrib-oadr-ven/pull/4/files#diff-0
> >
> > (2)
> > - *A* example-flows/demoflow_pullrequest.json
> > <
> https://github.com/Argonne-National-Laboratory/node-red-contrib-oadr-ven/pull/4/files#diff-1
> >
> > (1)
> > - *M* oadr2/oadr2-ven.js
> > <
> https://github.com/Argonne-National-Laboratory/node-red-contrib-oadr-ven/pull/4/files#diff-2
> >
> > (977)
> > - *A* oadr2/oadr2b-payload.js
> > <
> https://github.com/Argonne-National-Laboratory/node-red-contrib-oadr-ven/pull/4/files#diff-3
> >
> > (358)
> > - *A* oadr2/oadr2b_model/oadr2b_model.js
> > <
> https://github.com/Argonne-National-Laboratory/node-red-contrib-oadr-ven/pull/4/files#diff-4
> >
> > (6927)
> > - *A* oadr2/oadr2b_model/oadr2b_model.jsonschema
> > <
> https://github.com/Argonne-National-Laboratory/node-red-contrib-oadr-ven/pull/4/files#diff-5
> >
> > (22637)
> > - *D* package-lock.json
> > <
> https://github.com/Argonne-National-Laboratory/node-red-contrib-oadr-ven/pull/4/files#diff-6
> >
> > (1696)
> > - *M* package.json
> > <
> https://github.com/Argonne-National-Laboratory/node-red-contrib-oadr-ven/pull/4/files#diff-7
> >
> > (7)
> >
> > Patch Links:
> >
> > -
> >
> https://github.com/Argonne-National-Laboratory/node-red-contrib-oadr-ven/pull/4.patch
> > -
> >
> https://github.com/Argonne-National-Laboratory/node-red-contrib-oadr-ven/pull/4.diff
> >
> > —
> > You are receiving this because you are subscribed to this thread.
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/Argonne-National-Laboratory/node-red-contrib-oadr-ven/pull/4
> >,
> > or mute the thread
> > <
> https://github.com/notifications/unsubscribe-auth/ABsML2VqFXFvTysi9AreKoxjlPHg7bKvks5vWSPPgaJpZM4btoqb
> >
> > .
> >
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Argonne-National-Laboratory/node-red-contrib-oadr-ven/pull/4#issuecomment-474115223>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AASxPT3jNCP-ajWb8CjdqkeDf_mWqHPWks5vYAoNgaJpZM4btoqb>
> .
>

bnystrom commented:
I see what you are saying about the need for a second call to
model_builder....
Go ahead and make a pull request against Pull/No4 and I'll merge it for
further testing.

Best Regards
Bryan Nystrom


On Tue, Mar 19, 2019 at 10:37 AM zanni bertrand <notifications@github.com>
wrote:

> I merged your code into my fork and changed few things:
>
> - I had to uncomment the call to model_builder in
> oadrCreatePartyRegistration because of xmlSignature config. Indeed the only
> way to configure xmlSignature profile is by adding "xmlSignature:true" to
> createPartyRegistration payload. the line:
> xmlSignature = config.xmlSignature || false;
> at node-red module initialization is useless because there is no
> xmlSignature configuration at node level (node-red UI).
>
> - I added a proper oadr response when a oadrDistributeEvent payload is
> pushed from VTN to VEN
>
> - I changed the createdEvent function to support pull/push profile
>
> - I changed "responseCode" default value in different payload because this
> variable is supposed to be a string
>
> - I removed an unused dependency from package.json (node-webcrypto-p11)
>
> - I executed eslint over the code using this command: eslint ./oadr2/ --fix
> which seems to have done what I expected (code indent etc...). However I'm
> not used to eslint so don't hesitate to specify the workflow you prefer
> regarding that point. Maybe this should be scripted as "npm run" cmd in
> package.json file.
>
> Do you want me to make a pull request of those changes into the develop
> branch of your repo ?
>
> Best regards,
>
> Bertrand
>
> Le lun. 18 mars 2019 à 22:50, Bryan Nystrom <notifications@github.com> a
> écrit :
>
> > I've pulled your code into branch "pull/No4" and done some changes:
> >
> >
> > - Re-format and make cosmetic changes to satisfy the linter (please lint)
> > - Messages w/o XML signature default to adding the
> > oadrPayload/oadrSignedObject env wrappers
> > - Added an option in the config dialog to NOT do the above (in case
> > that's the way a particular VTN likes it)
> > - Always remove the "TYPE_NAME" property that is only needed by jsonix
> > transforms when outputing to Node-Red flows.
> > - Removed an extraneous extra call to the model_builder which existed in
> > the CreatePartyRegistraion function (I left it commented out in case you
> > have a valid reason for keeping it that I'm not aware of)
> >
> > Go ahead verify with that branch if I've broken anything on your end.
> I'll
> > do a bit more testing and if all is well, I'll release soon.
> >
> >
> > Best Regards
> > Bryan Nystrom
> >
> >
> > On Wed, Mar 13, 2019 at 11:14 AM zanni bertrand <
> notifications@github.com>
> > wrote:
> >
> > > Bryan,
> > >
> > > Here are the changes we discussed before. This pull request contains:
> > >
> > > - xmlSignature verify
> > > - compatibility with ECDSA P-384 keys
> > > - CreateReport / CancelReport queries
> > > - PartyReRegistrer / CancelPartyRegistration query
> > >
> > > I implemented the verify function the same way I implemented the sign
> > > function. Therefore, I had to "promisify" the message reading logic
> > > (prepareResMsg, prepareReqMsg) a bit.
> > >
> > > I also implemented functions to manipulate signatureWrapper you can use
> > > (hasWrapper, getWrappedPayloadName, wrap, unwrap). Have a look at
> > > prepareResMsg, prepareReqMsg in order to see how to use them.
> > >
> > > I modified a bit your demo flow in order to implements oadr business
> > logic
> > > on previously mentioned queries. You can found this modified flow in
> > > example-flows/demoflow_pullrequest.json file.
> > >
> > > Best regards
> > > Bertrand Zanni
> > > ------------------------------
> > > You can view, comment on, or merge this pull request online at:
> > >
> > >
> > >
> >
> https://github.com/Argonne-National-Laboratory/node-red-contrib-oadr-ven/pull/4
> > > Commit Summary
> > >
> > > - oadr20b model generation using jsonix
> > > - cleanup
> > > - xmldsig
> > > - xmldsig
> > > - xmldsig
> > > - cleanup + demo flow
> > > - profil flow variable
> > > - createOpt fix
> > > - xmldsig
> > > - canceledPartyRegistration response
> > > - handle createdReport canceledReport
> > > - xmldsig verify + xmldsig ECDSA P-384 + CreateReport + CancelReport +
> > > ReRegistration
> > >
> > > File Changes
> > >
> > > - *M* .gitignore
> > > <
> >
> https://github.com/Argonne-National-Laboratory/node-red-contrib-oadr-ven/pull/4/files#diff-0
> > >
> > > (2)
> > > - *A* example-flows/demoflow_pullrequest.json
> > > <
> >
> https://github.com/Argonne-National-Laboratory/node-red-contrib-oadr-ven/pull/4/files#diff-1
> > >
> > > (1)
> > > - *M* oadr2/oadr2-ven.js
> > > <
> >
> https://github.com/Argonne-National-Laboratory/node-red-contrib-oadr-ven/pull/4/files#diff-2
> > >
> > > (977)
> > > - *A* oadr2/oadr2b-payload.js
> > > <
> >
> https://github.com/Argonne-National-Laboratory/node-red-contrib-oadr-ven/pull/4/files#diff-3
> > >
> > > (358)
> > > - *A* oadr2/oadr2b_model/oadr2b_model.js
> > > <
> >
> https://github.com/Argonne-National-Laboratory/node-red-contrib-oadr-ven/pull/4/files#diff-4
> > >
> > > (6927)
> > > - *A* oadr2/oadr2b_model/oadr2b_model.jsonschema
> > > <
> >
> https://github.com/Argonne-National-Laboratory/node-red-contrib-oadr-ven/pull/4/files#diff-5
> > >
> > > (22637)
> > > - *D* package-lock.json
> > > <
> >
> https://github.com/Argonne-National-Laboratory/node-red-contrib-oadr-ven/pull/4/files#diff-6
> > >
> > > (1696)
> > > - *M* package.json
> > > <
> >
> https://github.com/Argonne-National-Laboratory/node-red-contrib-oadr-ven/pull/4/files#diff-7
> > >
> > > (7)
> > >
> > > Patch Links:
> > >
> > > -
> > >
> >
> https://github.com/Argonne-National-Laboratory/node-red-contrib-oadr-ven/pull/4.patch
> > > -
> > >
> >
> https://github.com/Argonne-National-Laboratory/node-red-contrib-oadr-ven/pull/4.diff
> > >
> > > —
> > > You are receiving this because you are subscribed to this thread.
> > > Reply to this email directly, view it on GitHub
> > > <
> >
> https://github.com/Argonne-National-Laboratory/node-red-contrib-oadr-ven/pull/4
> > >,
> > > or mute the thread
> > > <
> >
> https://github.com/notifications/unsubscribe-auth/ABsML2VqFXFvTysi9AreKoxjlPHg7bKvks5vWSPPgaJpZM4btoqb
> > >
> > > .
> > >
> >
> > —
> > You are receiving this because you authored the thread.
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/Argonne-National-Laboratory/node-red-contrib-oadr-ven/pull/4#issuecomment-474115223
> >,
> > or mute the thread
> > <
> https://github.com/notifications/unsubscribe-auth/AASxPT3jNCP-ajWb8CjdqkeDf_mWqHPWks5vYAoNgaJpZM4btoqb
> >
> > .
> >
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/Argonne-National-Laboratory/node-red-contrib-oadr-ven/pull/4#issuecomment-474429173>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABsMLztwBJw35hsEtg4JLkInSvBS0NiYks5vYQQfgaJpZM4btoqb>
> .
>
